### ISSUE TITLE
Use RolesAllowed with UserRights instead of UserRoles [5]

### ISSUE THREAD

**Issue by MartinWahnschaffe** [hzi-braunschweig/SORMAS-Project#5056]
### Problem Description
Hard-coded user roles need to be replaced with configurable ones. See #898.

Related to #2804, #2805 and #7820. Should be possible to work on this in parallel.

We need to be able to easily check whether the user has the needed permission to do something. One way to achieve this is to go on using the javax security RolesAllowed annotation, but instead of using it with UserRole, use it with UserRight.

### Proposed Change
The goal is to replace all usages of UserRole in javax security RolesAllowed with user rights.

- [x] Add SORMAS_REST user right as a replacement for the corresponding user role and the manual "USER" role. Give the user right to all user roles that need to access the REST api (users on district level and below for the mobile app, technical user roles that have to access the REST api)
- [x] Don't check the SORMAS_REST user right in every single resource anymore. Instead put it into the two identity stores.
- [x] Add SORMAS_TO_SORMAS_CLIENT user right ...
- [x] For REST_EXTERNAL_VISIT_USER introduce EXTERNAL_VISITS user right. Check for ExternalVisitsResource
- [x] For BAG_USER use the existing BAG_EXPORT user right
- [x] Add static final strings for all user rights to the UserRight enum (see UserRole) for usage with the RolesAllowed annotation
- [x] Move _SYSTEM string constant from UserRole to UserRight (not enum value)
- [ ] Write a code generator to update the mappings in web.xml and glassfish-web.xml for sormas-ui and sormas-rest. This should include all user rights
- [x] Update KeycloakIdentityStore, when it's properly configured and SormasIdentityStore to pass in the effective user rights instead of user roles to the realm
- [x] Replace all usages of UserRole in RolesAllowed annotations with the corresponding UserRight

Out of scope: It's not needed to add RolesAllowed annotations here, where there currently aren't any annotations yet. This will go into a separate issue.

**Comment by MartinWahnschaffe** [comment]
To be considered (thanks @StefanKock)

- RolesAllowed does not work with BeanTest - unit tests, insufficient permissions are not detectable with it.
- @HolgerReiseVSys  has built some auxiliary logic for another project that generates code snippets, e.g. for the XML files.
- Do not include string constants in the UserRight enum class (sormas-api), but as generated code in sormas-backend: Permission.SYSTEM ..., generator logic (incl. deviations from UserRight ), unit test that checks synchronicity.
- Consider defining default permissions on EJB classes, and restrict on certain methods -> may not be possible
- Non-intuitive behaviour with RolesAllowed on superclasses must be considered. Example: PersonService.delete must be overwritten and defined via RolesAllowed, otherwise the annotation of the superclass (to method or class) applies.
- Plan at which classes/layers RolesAllowed is defined restrictively (Facades) or generously (Services).
- Consider keeping a USER as a special permission (at least authenticated access).
- All in all, there is a great danger that the introduction of new RolesAllowed restrictions will prevent use cases in method cascades that have not been considered.

**Comment by MartinWahnschaffe** [comment]
Note: An alternative (or complementary) approach is to use [XACML (eXtensible Access Control Markup Language)](https://www.oasis-open.org/committees/download.php/2713/Brief_Introduction_to_XACML.html). This would be quite powerful, but would take more effort. It would likely be a good fit for attribute based access rights (in contrast to role based access rights, as needed here), although performance and maintenance implications would have to be evaluated first.

For the current use case a role based access control (RBAC) is needed and javax security seems to be the better fit.

**Comment by JonasCir** [comment]
@MartinWahnschaffe There is also this: https://docs.osohq.com/java/getting-started/quickstart.html
This allows to to write verifiably access policies as resource files. I have not used it but I'm impressed by their work :)

**Comment by JonasCir** [comment]
Hi I took a look at oso and coming from a (usable) security perspective it is just awesome.

**Comment by Scotchomani** [comment]
Hi, do you plan to implement this feature in the upcoming Sormas release ? https://app.moqups.com/3NUENEwuOE/view/page/a46f29001

**Comment by MateStrysewske** [comment]
@Scotchomani I think it's not planned for the upcoming release if I'm not mistaken

**Comment by MartinWahnschaffe** [comment]
I talked with Jonas about oso and decided to go with the current javax.security based approach. 
If there is a need for attribute based access control or more complex rules for RBAC we should consider oso and XACML.

**Comment by vidi42** [comment]
This might be blocked by #7820 because using `@RolesAllowed` will check the `KeycloakAccount.getCallerGroups` to see if the user has a certain role and this will only return the roles from the AccessToken. This is a problem because we want to remove the duplication of roles between SORMAS and Keycloak and only use those from SORMAS DB.
